File: src/mtc_pipeline/src/move_to_stages.cpp
#include "mtc_pipeline/move_to_stages.hpp"
#include <moveit/task_constructor/stages/move_to.h>
#include <nlohmann/json.hpp>

MoveToStages::MoveToStages(const rclcpp::Node::SharedPtr& node)
    : BaseStages(node) {}

bool MoveToStages::run(const mtc_pipeline::action::MoveToAction::Goal& goal) {
    auto task = create_task_template("MoveTo Task");
    auto planner = (goal.planning_type == "cartesian")
        ? make_cartesian_planner() : make_pipeline_planner();

    // Relative move (direction + distance)
    if (!goal.direction.empty() && goal.distance != 0.0) {
        auto label = "move_" + goal.direction + "_" + std::to_string(goal.distance) + "m";
        auto stage = create_relative_move_stage(label, goal.direction, goal.distance, planner);
        stage->properties().configureInitFrom(mtc::Stage::PARENT, {"group", "ik_frame"});
        task.add(std::move(stage));
    }
    // Target-based move (joint pose or named state)
    else if (!goal.target.empty()) {
        nlohmann::json poses;
        try {
            poses = nlohmann::json::parse(goal.poses_json);
        } catch (const nlohmann::json::exception& e) {
            RCLCPP_ERROR(node()->get_logger(), "Failed to parse poses_json: %s", e.what());
            return false;
        }

        auto stage = std::make_unique<mtc::stages::MoveTo>("move_to_" + goal.target, planner);
        stage->properties().configureInitFrom(mtc::Stage::PARENT, {"group", "ik_frame"});
        stage->setGroup(default_arm_group_name());

        if (poses.contains(goal.target)) {
            // Joint pose from JSON
            const auto& joint_pose = poses.at(goal.target);
            if (!joint_pose.is_array() || joint_pose.size() != 6) {
                RCLCPP_ERROR(node()->get_logger(), "'%s' must be array of 6 joint angles", goal.target.c_str());
                return false;
            }
            stage->setGoal(joints_from_degrees(joint_pose.get<std::vector<double>>()));
        } else {
            // Named SRDF state
            stage->setGoal(goal.target);
        }
        task.add(std::move(stage));
    }
    else {
        RCLCPP_ERROR(node()->get_logger(), "MoveTo: missing 'target' or 'direction'/'distance'");
        return false;
    }

    return load_plan_execute(task);
}

File: src/mtc_py/mtc_py_lib/stages/move_to_stages.py
"""MoveTo stages - Python equivalent of move_to_stages.hpp/cpp.

Handles MoveTo operations:
- Relative moves (direction + distance)
- Target-based moves (joint poses from JSON or named SRDF states)
"""

import json
from typing import Dict, Any
from moveit.task_constructor import core, stages
from mtc_py_lib.stages.base_stages import BaseStages


class MoveToStages(BaseStages):
    """Handles MoveTo action: relative moves, joint poses, named states."""

    def run(self, goal) -> bool:
        """Execute MoveTo action.

        Args:
            goal: MoveToAction.Goal with fields:
                - target: Pose name, SRDF state, or empty for relative moves
                - planning_type: "joint" or "cartesian"
                - direction: Direction for relative moves
                - distance: Distance in meters for relative moves
                - poses_json: JSON string with pose definitions

        Returns:
            True if successful, False otherwise
        """
        task = self.create_task_template("MoveTo Task")

        # Select planner based on planning_type
        planning_type = goal.planning_type if goal.planning_type else "joint"
        if planning_type == "cartesian":
            planner = self.make_cartesian_planner()
            self.logger.info("Using Cartesian planner")
        else:
            planner = self.make_pipeline_planner()
            self.logger.info("Using pipeline planner")

        # Case 1: Relative move (direction + distance)
        if goal.direction and goal.distance != 0.0:
            stage = self.create_relative_move_stage(
                f"move_{goal.direction}_{goal.distance:.3f}m",
                goal.direction,
                goal.distance,
                planner
            )
            task.add(stage)
            self.logger.info(
                f"Planning relative move: {goal.direction} {goal.distance}m"
            )

        # Case 2: Target-based move
        elif goal.target:
            poses = self._parse_poses(goal.poses_json)

            move_stage = stages.MoveTo(f"move_to_{goal.target}", planner)
            move_stage.group = self.arm_group

            # Check if target is a defined joint pose in the JSON
            if goal.target in poses:
                joint_values = poses[goal.target]
                if isinstance(joint_values, list):
                    move_stage.setGoal(self.joints_from_degrees(joint_values))
                    self.logger.info(f"Planning move to joint pose: {goal.target}")
                else:
                    self.logger.error(
                        f"Invalid pose format for '{goal.target}': "
                        f"expected list, got {type(joint_values)}"
                    )
                    return False
            else:
                # Assume it's a named SRDF state
                move_stage.setGoal(goal.target)
                self.logger.info(f"Planning move to named state: {goal.target}")

            task.add(move_stage)

        else:
            self.logger.error(
                "No valid move target specified. "
                "Provide either (direction + distance) or target."
            )
            return False

        return self.load_plan_execute(task)

    def _parse_poses(self, poses_json: str) -> Dict[str, Any]:
        """Parse poses JSON string.

        Args:
            poses_json: JSON string containing pose definitions

        Returns:
            Dictionary of pose names to values
        """
        if not poses_json:
            return {}

        try:
            return json.loads(poses_json)
        except json.JSONDecodeError as e:
            self.logger.error(f"Failed to parse poses JSON: {e}")
            return {}
